src/libpmem/pmem.c

```diff
@@ -530,6 +564,279 @@ pmem_is_pmem(const void *addr, size_t len)
 	return Func_is_pmem(addr, len);
 }
 
+
+/*
+ * map_range_comparer -- (internal) compares the two mapping trackers
+ */
+static intptr_t
+map_range_comparer(struct map_tracker *a, struct map_tracker *b)
+{
+	return ((intptr_t)a->base_addr - (intptr_t)b->base_addr);
+}
+
+/*
+ * map_range_find --(internal) find the map tracker for given address range
+ *
+ * Returns the first entry at least partially overlapping given range.
+ * It's up to the caller to check whether the entry exactly matches the range,
+ * or if the range spans multiple entries.
+ * The caller is also responsible for acquiring/releasing a lock on
+ * the map tracking list.
+ */
+static struct map_tracker *
+map_range_find(uintptr_t addr, size_t len)
+{
+	LOG(10, "addr 0x%016" PRIxPTR " len %zu", addr, len);
+
+	uintptr_t end = addr + len;
+
+	struct map_tracker *mt;
+	SORTEDQ_FOREACH(mt, &Mmap_list, entry) {
+		if (addr < mt->end_addr &&
+		    (addr >= mt->base_addr || end > mt->base_addr))
+			return mt;
+
+		/* break if there is no chance to find matching entry */
+		if (addr < mt->base_addr)
+			break;
+	}
+
+	return NULL;
+}
+
+/*
+ * map_range_register -- (internal) add a memory range into a map tracking list
+ */
+int
+map_range_register(const void *addr, size_t len, int fd)
+{
+	LOG(3, "addr %p len %zu fd %d", addr, len, fd);
+
+	int ret = 0;
+
+	ASSERT(fd >= 0);
+
+	if (os_rwlock_wrlock(&Mmap_list_lock)) {
+		errno = EBUSY;
```

Comment:
os_rwlock_rdlock returns an errno.

Reply:
I suggest to write util_ variants of rwlocks if they are needed. I never liked error checking around locks.